watcherx/changefeed.go

```diff
@@ -68,7 +68,10 @@ const heartBeatInterval = time.Second
 // Watcher.DispatchNow() does not have an effect in this method.
 //
 // This watcher is blocking to allow proper context cancellation and clean up.
-func WatchChangeFeed(ctx context.Context, cx *sqlx.DB, tableName string, c EventChannel, cursor time.Time) (_ Watcher, err error) {
+func WatchChangeFeed(ctx context.Context, cx *sqlx.DB, tableName string, out EventChannel, cursor time.Time) (_ Watcher, err error) {
+	c := make(EventChannel)
+	deduplicate(c, out, 100)
```

Comment:
How do you determine this (the 100 mark)? Would it be better to instead set a time (e.g. 15 seconds)?

Reply:
Ristretto is an eventually consistent cache that has no guarantee that writes end up in the cache! For de-duplication, we would indeed need something more consistent like the ring buffer used here. There are projects like https://github.com/patrickmn/go-cache but we should try to keep the memory consumption of the cache, even on high load, to < 50mb

Reply:
The idea behind the ring buffer with the fixed number of elements was that we want constant runtime (as it sits between the changefeed and the consumers) and constant memory consumption. If we evict not based on the total elements in the set/cache, but based on time, the data structure must support an arbitrary amount of elements.

I still think the impementation with the constant number of elements is fine, especially since the previous behaviour was that there was no deduplication at all. If a time-based eviction policy is desired, the eviction probably needs to happen in a separate goroutine to not block forwarding the events. In any case I'd rather implement this for this special use case than use a general-purpose datastructure.

Reply:
SGTM